### ISSUE TITLE
Failed to get disk format of disk /dev/nvme2n1: exit status 127 eks

### ISSUE THREAD

**Issue by darklight147** [bottlerocket-os/bottlerocket#3556]
**Image I'm using:**

BOTTLEROCKET_ARM_64 on AWS EKS 1.27

**What I expected to happen:**

Volumes Mounted onto the pods

**What actually happened:**

Pods crash and Volumes don't mount onto the pods even though the volumes are well provisioned

**How to reproduce the problem:**

AWS EKS cluster version 1.27 with Nodegroup BOTTLEROCKET_ARM_64  a1.xlarge and addon `aws-ebs-csi-driver`

try running the sonarqube helm chart with persistance enabled and storageClass `gp2`

```sh
helm upgrade --install -n sonarqube sonarqube sonarqube/sonarqube --create-namespace
```

The pods won't start because the pod can't mount the volumes

check by running `kubectl describe po POD_NAME`

**Comment by darklight147** [comment]
PS: i do understand that the image `postgres` supports only amd64 arch, on which i set a nodeSelector to a different node which is amd64 based

**Comment by yeazelm** [comment]
Hello @darklight147, thanks for opening this issue! I wanted to ask a couple clarifying questions.

Does this work on `x86_64`? You called out that `postgres` needs `x86_64` but it would be good to know if this issue is limited to a particular architecture or if that actually isn't the problem. 

Can you share a few more details about "don't mount onto the pods even though the volumes are well provisioned"? It appears this helm chart needs some specific configuration and there are a lot of options: https://github.com/SonarSource/helm-chart-sonarqube/tree/master/charts/sonarqube so narrowing down what storage settings you've chosen would help us see what is going on. Can you share some additional data about how you have configured the sonarqube helm chart plus what volumes you have configured with the EBS CSI Driver? 

I haven't gotten to the same point in provisioning this chart due to the additional setup needed for storage, so until we can reproduce this ourselves, can you include a bit more logs/describes from your cluster to help us look at this issue?

**Comment by darklight147** [comment]
@yeazelm I have test with a botterocket x86_64 - only Cluster and the helm chart works just fine, running pods, and mounted volumes

as for the Helm chart i've pulled the latest chart, enabled `persistence.enabled` to `true` and `persistence.storageClass` to `gp2`

to reproduce you could run the following

```sh

helm pull sonarqube/sonarqube

tar xvf sonarqube-VERSION.tgz

cd sonarqube

# change the values i mentioned above then:

helm upgrade --install -n sonarqube sonarqube . --create-namespace
```

**Comment by darklight147** [comment]
[Odoo](https://bitnami.com/stack/odoo/helm) is also another Helm chart am struggling with running on arm64 botterocket since it also relies on postgresql 

```sh

➜  odoo k get po                           
NAME                    READY   STATUS              RESTARTS   AGE
odoo-6f89dc694b-p8s4q   0/1     ContainerCreating   0          2m8s
odoo-postgresql-0       0/1     ContainerCreating   0          2m8s
➜  odoo 
```


The description of the odoo-xxxx pod


```sh
➜  odoo k logs odoo-6f89dc694b-p8s4q                    
Error from server (BadRequest): container "odoo" in pod "odoo-6f89dc694b-p8s4q" is waiting to start: ContainerCreating
➜  odoo k describe po odoo-6f89dc694b-p8s4q                     
Name:             odoo-6f89dc694b-p8s4q
Namespace:        odoo
Priority:         0
Service Account:  odoo
Node:             ip-172-31-20-230.ec2.internal/172.31.20.230
Start Time:       Tue, 31 Oct 2023 22:40:24 +0100
Labels:           app.kubernetes.io/instance=odoo
                  app.kubernetes.io/managed-by=Helm
                  app.kubernetes.io/name=odoo
                  helm.sh/chart=odoo-23.0.1
                  pod-template-hash=6f89dc694b
Annotations:      <none>
Status:           Pending
IP:               
IPs:              <none>
Controlled By:    ReplicaSet/odoo-6f89dc694b
Containers:
  odoo:
    Container ID:   
    Image:          marketplace.azurecr.io/bitnami/odoo:16.0.20221015-debian-11-r0
    Image ID:       
    Port:           8069/TCP
    Host Port:      0/TCP
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Requests:
      cpu:      300m
      memory:   512Mi
➜  odoo k get po                           
NAME                    READY   STATUS              RESTARTS   AGE
odoo-6f89dc694b-p8s4q   0/1     ContainerCreating   0          2m8s
odoo-postgresql-0       0/1     ContainerCreating   0          2m8s
➜  odoo k describe po odoo-6f89dc694b-p8s4q
Name:             odoo-6f89dc694b-p8s4q
Namespace:        odoo
Priority:         0
Service Account:  odoo
Node:             ip-172-31-20-230.ec2.internal/172.31.20.230
Start Time:       Tue, 31 Oct 2023 22:40:24 +0100
Labels:           app.kubernetes.io/instance=odoo
                  app.kubernetes.io/managed-by=Helm
                  app.kubernetes.io/name=odoo
                  helm.sh/chart=odoo-23.0.1
                  pod-template-hash=6f89dc694b
Annotations:      <none>
Status:           Pending
IP:               
IPs:              <none>
Controlled By:    ReplicaSet/odoo-6f89dc694b
Containers:
  odoo:
    Container ID:   
    Image:          marketplace.azurecr.io/bitnami/odoo:16.0.20221015-debian-11-r0
    Image ID:       
    Port:           8069/TCP
    Host Port:      0/TCP
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Requests:
      cpu:      300m
      memory:   512Mi
    Liveness:   http-get http://:http/ delay=600s timeout=5s period=30s #success=1 #failure=6
    Readiness:  http-get http://:http/ delay=30s timeout=5s period=10s #success=1 #failure=6
    Environment:
      BITNAMI_DEBUG:              false
      ALLOW_EMPTY_PASSWORD:       no
      ODOO_DATABASE_HOST:         odoo-postgresql
      ODOO_DATABASE_PORT_NUMBER:  5432
      ODOO_DATABASE_NAME:         bitnami_odoo
      ODOO_DATABASE_USER:         bn_odoo
      ODOO_DATABASE_PASSWORD:     <set to the key 'password' in secret 'odoo-postgresql'>  Optional: false
      ODOO_EMAIL:                 user@example.com
      ODOO_PASSWORD:              <set to the key 'odoo-password' in secret 'odoo'>  Optional: false
      ODOO_SKIP_BOOTSTRAP:        no
      ODOO_LOAD_DEMO_DATA:        no
    Mounts:
      /bitnami/odoo from odoo-data (rw)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  odoo-data:
    Type:        PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:   odoo
    ReadOnly:    false
QoS Class:       Burstable
Node-Selectors:  odoo=true
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason                  Age                  From                     Message
  ----     ------                  ----                 ----                     -------
  Normal   Scheduled               2m36s                default-scheduler        Successfully assigned odoo/odoo-6f89dc694b-p8s4q to ip-172-31-20-230.ec2.internal
  Normal   SuccessfulAttachVolume  2m33s                attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-10872f02-b800-42a2-87e3-dd4ba23c0f97"
  Warning  FailedMount             33s                  kubelet                  Unable to attach or mount volumes: unmounted volumes=[odoo-data], unattached volumes=[], failed to process volumes=[]: timed out waiting for the condition
  Warning  FailedMount             24s (x9 over 2m32s)  kubelet                  MountVolume.MountDevice failed for volume "pvc-10872f02-b800-42a2-87e3-dd4ba23c0f97" : rpc error: code = Internal desc = could not format "/dev/nvme3n1" and mount it at "/var/lib/kubelet/plugins/kubernetes.io/csi/ebs.csi.aws.com/af78dca5f60e7543c95d814ea4ab673cac6f4cd2769e85ce45400095fe2d846f/globalmount": failed to get disk format of disk /dev/nvme3n1: exit status 127
➜  odoo 
```


```sh
➜  odoo k describe po odoo-postgresql-0    
Name:             odoo-postgresql-0
Namespace:        odoo
Priority:         0
Service Account:  default
Node:             ip-172-31-20-230.ec2.internal/172.31.20.230
Start Time:       Tue, 31 Oct 2023 22:40:24 +0100
Labels:           app.kubernetes.io/component=primary
                  app.kubernetes.io/instance=odoo
                  app.kubernetes.io/managed-by=Helm
                  app.kubernetes.io/name=postgresql
                  apps.kubernetes.io/pod-index=0
                  controller-revision-hash=odoo-postgresql-58489bb46d
                  helm.sh/chart=postgresql-12.0.0
                  statefulset.kubernetes.io/pod-name=odoo-postgresql-0
Annotations:      <none>
Status:           Pending
IP:               
IPs:              <none>
Controlled By:    StatefulSet/odoo-postgresql
Containers:
  postgresql:
    Container ID:   
    Image:          marketplace.azurecr.io/bitnami/postgresql:15.0.0-debian-11-r1
    Image ID:       
    Port:           5432/TCP
    Host Port:      0/TCP
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Requests:
      cpu:      250m
      memory:   256Mi
    Liveness:   exec [/bin/sh -c exec pg_isready -U "bn_odoo" -d "dbname=bitnami_odoo" -h 127.0.0.1 -p 5432] delay=30s timeout=5s period=10s #success=1 #failure=6
    Readiness:  exec [/bin/sh -c -e exec pg_isready -U "bn_odoo" -d "dbname=bitnami_odoo" -h 127.0.0.1 -p 5432
[ -f /opt/bitnami/postgresql/tmp/.initialized ] || [ -f /bitnami/postgresql/.initialized ]
] delay=5s timeout=5s period=10s #success=1 #failure=6
    Environment:
      BITNAMI_DEBUG:                        false
      POSTGRESQL_PORT_NUMBER:               5432
      POSTGRESQL_VOLUME_DIR:                /bitnami/postgresql
      PGDATA:                               /bitnami/postgresql/data
      POSTGRES_USER:                        bn_odoo
      POSTGRES_POSTGRES_PASSWORD:           <set to the key 'postgres-password' in secret 'odoo-postgresql'>  Optional: false
      POSTGRES_PASSWORD:                    <set to the key 'password' in secret 'odoo-postgresql'>           Optional: false
      POSTGRES_DB:                          bitnami_odoo
      POSTGRESQL_ENABLE_LDAP:               no
      POSTGRESQL_ENABLE_TLS:                no
      POSTGRESQL_LOG_HOSTNAME:              false
      POSTGRESQL_LOG_CONNECTIONS:           false
      POSTGRESQL_LOG_DISCONNECTIONS:        false
      POSTGRESQL_PGAUDIT_LOG_CATALOG:       off
      POSTGRESQL_CLIENT_MIN_MESSAGES:       error
      POSTGRESQL_SHARED_PRELOAD_LIBRARIES:  pgaudit
    Mounts:
      /bitnami/postgresql from data (rw)
      /dev/shm from dshm (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-p5h27 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  data-odoo-postgresql-0
    ReadOnly:   false
  dshm:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  <unset>
  kube-api-access-p5h27:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason                  Age                  From                     Message
  ----     ------                  ----                 ----                     -------
  Normal   Scheduled               3m17s                default-scheduler        Successfully assigned odoo/odoo-postgresql-0 to ip-172-31-20-230.ec2.internal
  Normal   SuccessfulAttachVolume  3m14s                attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-854fe613-5ebe-4724-b80c-c3fe4a5ca66b"
  Warning  FailedMount             74s                  kubelet                  Unable to attach or mount volumes: unmounted volumes=[data], unattached volumes=[], failed to process volumes=[]: timed out waiting for the condition
  Warning  FailedMount             65s (x9 over 3m13s)  kubelet                  MountVolume.MountDevice failed for volume "pvc-854fe613-5ebe-4724-b80c-c3fe4a5ca66b" : rpc error: code = Internal desc = could not format "/dev/nvme2n1" and mount it at "/var/lib/kubelet/plugins/kubernetes.io/csi/ebs.csi.aws.com/b0a3f86b78a57d74bc0f57053080a402bc41698e47adec93117167e84a84d06e/globalmount": failed to get disk format of disk /dev/nvme2n1: exit status 127
➜  odoo 
```

# The PVCs


```sh

➜  odoo k get pvc                      
NAME                     STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
data-odoo-postgresql-0   Bound    pvc-854fe613-5ebe-4724-b80c-c3fe4a5ca66b   8Gi        RWO            gp2            3m48s
odoo                     Bound    pvc-10872f02-b800-42a2-87e3-dd4ba23c0f97   10Gi       RWO            gp2            3m48s
➜  odoo 

```

Kubectl describe pvc

```sh
➜  odoo k describe pvc                       
Name:          data-odoo-postgresql-0
Namespace:     odoo
StorageClass:  gp2
Status:        Bound
Volume:        pvc-854fe613-5ebe-4724-b80c-c3fe4a5ca66b
Labels:        app.kubernetes.io/component=primary
               app.kubernetes.io/instance=odoo
               app.kubernetes.io/name=postgresql
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: ebs.csi.aws.com
               volume.kubernetes.io/selected-node: ip-172-31-20-230.ec2.internal
               volume.kubernetes.io/storage-provisioner: ebs.csi.aws.com
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      8Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Used By:       odoo-postgresql-0
Events:
  Type    Reason                 Age                 From                                                                                      Message
  ----    ------                 ----                ----                                                                                      -------
  Normal  WaitForFirstConsumer   4m                  persistentvolume-controller                                                               waiting for first consumer to be created before binding
  Normal  Provisioning           4m                  ebs.csi.aws.com_ebs-csi-controller-7d984fd55b-rwph7_bdbece79-2d22-4942-abd7-a9e2d375345c  External provisioner is provisioning volume for claim "odoo/data-odoo-postgresql-0"
  Normal  ExternalProvisioning   3m58s (x2 over 4m)  persistentvolume-controller                                                               Waiting for a volume to be created either by the external provisioner 'ebs.csi.aws.com' or manually by the system administrator. If volume creation is delayed, please verify that the provisioner is running and correctly registered.
  Normal  ProvisioningSucceeded  3m56s               ebs.csi.aws.com_ebs-csi-controller-7d984fd55b-rwph7_bdbece79-2d22-4942-abd7-a9e2d375345c  Successfully provisioned volume pvc-854fe613-5ebe-4724-b80c-c3fe4a5ca66b


Name:          odoo
Namespace:     odoo
StorageClass:  gp2
Status:        Bound
Volume:        pvc-10872f02-b800-42a2-87e3-dd4ba23c0f97
Labels:        app.kubernetes.io/instance=odoo
               app.kubernetes.io/managed-by=Helm
               app.kubernetes.io/name=odoo
               helm.sh/chart=odoo-23.0.1
Annotations:   meta.helm.sh/release-name: odoo
               meta.helm.sh/release-namespace: odoo
               pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: ebs.csi.aws.com
               volume.kubernetes.io/selected-node: ip-172-31-20-230.ec2.internal
               volume.kubernetes.io/storage-provisioner: ebs.csi.aws.com
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      10Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Used By:       odoo-6f89dc694b-p8s4q
Events:
  Type    Reason                 Age                 From                                                                                      Message
  ----    ------                 ----                ----                                                                                      -------
  Normal  WaitForFirstConsumer   4m                  persistentvolume-controller                                                               waiting for first consumer to be created before binding
  Normal  Provisioning           4m                  ebs.csi.aws.com_ebs-csi-controller-7d984fd55b-rwph7_bdbece79-2d22-4942-abd7-a9e2d375345c  External provisioner is provisioning volume for claim "odoo/odoo"
  Normal  ExternalProvisioning   3m58s (x2 over 4m)  persistentvolume-controller                                                               Waiting for a volume to be created either by the external provisioner 'ebs.csi.aws.com' or manually by the system administrator. If volume creation is delayed, please verify that the provisioner is running and correctly registered.
  Normal  ProvisioningSucceeded  3m56s               ebs.csi.aws.com_ebs-csi-controller-7d984fd55b-rwph7_bdbece79-2d22-4942-abd7-a9e2d375345c  Successfully provisioned volume pvc-10872f02-b800-42a2-87e3-dd4ba23c0f97
```

@yeazelm

**Comment by darklight147** [comment]
Update: 

After running cordon on the arm64 node, and leaving only the amd64 botterocket node, the pods are running in the odoo helm chart

```sh
➜  odoo k get no --selector kubernetes.io/arch=arm64 | awk 'NR!=1 {print $1}' | xargs kubectl cordon
node/ip-172-31-20-230.ec2.internal cordoned
➜  odoo helm upgrade --install -n odoo odoo . --create-namespace
Release "odoo" does not exist. Installing it now.
NAME: odoo
LAST DEPLOYED: Tue Oct 31 22:55:43 2023
NAMESPACE: odoo
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
CHART NAME: odoo
CHART VERSION: 23.0.1
APP VERSION: 16.0.20221015

** Please be patient while the chart is being deployed **

1. Get the Odoo URL indicated on the Ingress Rule and associate it to your cluster external IP:

   export CLUSTER_IP=$(minikube ip) # On Minikube. Use: `kubectl cluster-info` on others K8s clusters
   export HOSTNAME=$(kubectl get ingress --namespace odoo odoo -o jsonpath='{.spec.rules[0].host}')
   echo "Odoo URL: http://$HOSTNAME/"
   echo "$CLUSTER_IP  $HOSTNAME" | sudo tee -a /etc/hosts

2. Obtain the login credentials

  export ODOO_EMAIL=user@example.com
  export ODOO_PASSWORD=$(kubectl get secret --namespace "odoo" odoo -o jsonpath="{.data.odoo-password}" | base64 -d)

  echo Email   : $ODOO_EMAIL
  echo Password: $ODOO_PASSWORD
➜  odoo kgp                                                                                         
NAME                    READY   STATUS              RESTARTS   AGE
odoo-6f89dc694b-kbvjg   0/1     ContainerCreating   0          6s
odoo-postgresql-0       0/1     ContainerCreating   0          6s
➜  odoo kgpw
NAME                    READY   STATUS              RESTARTS   AGE
odoo-6f89dc694b-kbvjg   0/1     ContainerCreating   0          8s
odoo-postgresql-0       0/1     ContainerCreating   0          8s
odoo-6f89dc694b-kbvjg   0/1     Running             0          10s
odoo-postgresql-0       0/1     Running             0          13s
odoo-postgresql-0       1/1     Running             0          23s
```


```sh
➜  odoo kgno -owide
NAME                            STATUS                     ROLES    AGE   VERSION               INTERNAL-IP     EXTERNAL-IP     OS-IMAGE                                KERNEL-VERSION   CONTAINER-RUNTIME
ip-172-31-16-128.ec2.internal   Ready                      <none>   17d   v1.27.4-eks-cedffd4   172.31.16.128   MASKED   Bottlerocket OS 1.15.1 (aws-k8s-1.27)   5.15.128         containerd://1.6.23+bottlerocket
ip-172-31-20-230.ec2.internal   Ready,SchedulingDisabled   <none>   8d    v1.27.4-eks-cedffd4   172.31.20.230   MASKED   Bottlerocket OS 1.15.1 (aws-k8s-1.27)   5.15.128         containerd://1.6.23+bottlerocket
➜  odoo 
```

**Comment by yeazelm** [comment]
Thanks for all the great info! This really helps. I noticed `rpc error: code = Internal desc = could not format "/dev/nvme2n1" and mount it at ` and `failed to get disk format of disk /dev/nvme2n1: exit status 127` which is curious for sure. Do you happen to know what filesystem these disks should be? I'm not sure if you specified one in your configs, if not, we'll need to dig into the CSI Driver on what its doing. It might be useful to compare the disk info from the `x86_64` box that is working with the `arm64` just to see. 

This should help me focus in on what might be the issue, I'll take a deeper look and get back to you on this ticket.

**Comment by darklight147** [comment]
@yeazelm Here is the output from one of the pods that have succefully run in bottlerocket x86_64

```bash
# df -Th
Filesystem     Type     Size  Used Avail Use% Mounted on
overlay        overlay   20G   14G  6.6G  67% /
tmpfs          tmpfs     64M     0   64M   0% /dev
/dev/xvdaa     ext4     9.8G  2.8M  9.8G   1% /bitnami/odoo
/dev/xvdb1     ext4      20G   14G  6.6G  67% /etc/hosts
shm            tmpfs     64M     0   64M   0% /dev/shm
tmpfs          tmpfs    3.9G     0  3.9G   0% /proc/acpi
tmpfs          tmpfs    3.9G     0  3.9G   0% /proc/scsi
tmpfs          tmpfs    3.9G     0  3.9G   0% /sys/firmware
# 
```

And from the postgresql pod

```sh
$ df -Th
Filesystem     Type     Size  Used Avail Use% Mounted on
overlay        overlay   20G   14G  6.6G  67% /
tmpfs          tmpfs     64M     0   64M   0% /dev
/dev/xvdb1     ext4      20G   14G  6.6G  67% /etc/hosts
tmpfs          tmpfs    7.1G  1.1M  7.1G   1% /dev/shm
/dev/xvdab     ext4     7.8G   74M  7.7G   1% /bitnami/postgresql
tmpfs          tmpfs    7.1G   12K  7.1G   1% /run/secrets/kubernetes.io/serviceaccount
tmpfs          tmpfs    3.9G     0  3.9G   0% /proc/acpi
tmpfs          tmpfs    3.9G     0  3.9G   0% /proc/scsi
tmpfs          tmpfs    3.9G     0  3.9G   0% /sys/firmware
$ 

```

**Comment by yeazelm** [comment]
Thanks for the update, in my searching for what might be wrong, I did find https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/1784 - can you confirm which version of the EBS CSI Driver you are using? This could be part of the issue.

**Comment by darklight147** [comment]
@yeazelm Sorry for the delay, yes indeed it is `v1.24.0-eksbuild.1`

**Comment by webern** [comment]
The CSI driver issue says that it has been fixed in `v1.24.1+` and you have reported using `v1.24.0`. Can you try updating the CSI driver? https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/1784

**Comment by darklight147** [comment]
Fixed on v1.24.0
